### PR TITLE
Improve font loading strategy

### PR DESCRIPTION
**Pull request by Nooshu**
## Issue
This PR has been raised to solve one of the major frontend performance issues that is currently live on GOV.UK. Four WOFF compressed web-fonts are base64 encoded and are sitting on the browsers critical path. This 200KB CSS file must be downloaded and parsed before page rendering can start. This single download is approximately 82% of the render blocking assets which stop the browser showing any content to the user.

Base64 encoding into a CSS file is now [considered an anti-pattern](https://csswizardry.com/2017/02/base64-encoding-and-performance/#fn:1) so the font loading method is being revisited.

## Solution
This PR removes the Base64 encoded font and reverts to a standard HTTP request to load the font files. This gives us the added advantage of being able to offer multiple font sources. The browser can then choose the most appropriate supported font format.

The fonts being used are unchanged. We have created a new WOFF2 version of the font which offers approximately 25% better compression and no change in quality.

This PR is also taking advantage of a new CSS property that is [available in modern browsers](https://caniuse.com/#search=font-display): `font-display`. This property defines how the browser behaves during the font download. Browsers that don't support this property will ignore it. More information about `font-display` can be seen in [this article](https://css-tricks.com/almanac/properties/f/font-display/).

## Font display simulated testing
I simulated a 2G connection and a mobile device and used [Sitespeed.io](https://www.sitespeed.io/) to compare the visual rendering differences between the font-display values. Each video was spliced into the following order from left to right:

1. Current live setup (base)
2. Standard font-face, WOFF2 and font-display block
3. Standard font-face, WOFF2 and font-display swap
4. Standard font-face, WOFF2 and font-display fallback
5. Standard font-face, WOFF2 and font-display optional

Each image below shows the page render of each method tested above.

vis1
_At 183ms the font-display optional (5) starts to render the page. Font has already been loaded, just waiting on the icon to load. Non of the other tests have started painting yet._

vis2
_At 200ms font-display swap paints the page. Font loaded but still missing icons and logo._

vis3
_At 217ms font-display fallback (4) paints the page. JavaScript on optional (5) has executed as can be seen by the visible cookie banner. Note that the base (1) and font-display block have yet to paint the page._

vis4
_At 233ms base (1) starts to paint the page. Note the Flash Of Invisible Text (FOIT) that can be seen in this screenshot due to the font still being loaded._

vis5
_At 300ms most of the font-display paints are complete apart from minor icon / logo loading not related to the font loading technique. The base (1) is still incomplete and no content can be read._

vis6
_At 433ms all video paints are complete. Base (1) completed at 350ms._

As you can see above the new font loading method and any of the font-display properties beat our current loading method on a simulated 2G connection.

For the moment we are using `font-display: fallback`. `fallback` acts as a compromise between the `auto` and `swap` values, so it is a good baseline we can iterate from in the future. This will show the fallback font (Arial) after 100ms if NTA hasn't been downloaded.  The browser will only swap to NTA if it is still within the short swap period (~3 seconds). Other display methods will be considered in the future once tested on the live environment. The staging environment connection was too slow which caused `optional` to fall back to Arial in many cases. 

## Speedcurve Results 
The staging environment has been used to compare the old and new font loading method on the GOV.UK homepage. The image below has been generated by SpeedCurve. Percentage differences between new and old are displayed in the middle.

speedcurve

Total page size has been reduced by 16%, and an 82% reduction in the size of the CSS used on the page. All rendering metrics show large percentage increase meaning the browser renders the page quicker using the new loading method.

### Fonts
More information about the fonts being used and the filename convention can be found in this [private repo](https://github.com/alphagov/fonts).

### CODE REVIEW

**Comment by nickcolley** [comment]
Hey Matt,

Normally we would bump the changelog with an unreleased heading, then do an additional pull request to do the release. See https://github.com/alphagov/govuk_template/pull/327/files for an example of this.

File: source/assets/stylesheets/fonts/filename-generator.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+filename=${1?Usage: $0 <filename>}
+
+md5hash=$( md5 -q "$filename" | cut -c -10 )
+filename_ext=${filename##*.}
+
+echo -n "What is the version number? [default: 1] "
+read -r version
+: "${version:=1}"
+
+echo -n "What type of font (e.g. tabular numbers)? [default: ''] "
```
Comment (nickcolley) [inline source/assets/stylesheets/fonts/filename-generator.sh]: can the examples by only what you should type?

It looks like you'd only want to type 'tabular' here? not 'tabular numbers'?

File: source/assets/stylesheets/fonts/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Naming convention
```
[the remark this answers is not in the document]
Comment (Nooshu) [reply source/assets/stylesheets/fonts/README.md]: Thanks, fixed!

File: source/assets/stylesheets/fonts/filename-generator.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#/bin/bash
```
[the remark this answers is not in the document]
Comment (Nooshu) [reply source/assets/stylesheets/fonts/filename-generator.sh]: Thanks, fixed!

**Comment by 36degrees** [comment]
Sorry, one more thing - can we squash the commits together and/or make sure the first commit doesn't reference the version number bump – that'll cause confusion for anyone looking through the history later as we're now not bumping the version number as part of that commit.

**Comment by tijmenb** [comment]
@36degrees that's there to be able to test the gem on static as a workaround for https://github.com/alphagov/static/pull/1325#issuecomment-372307499, in the actual PR we don't want `/app` committed.

**Comment by steventux** [comment]
@tijmenb @nickcolley I've cleaned this PR branch up - are you happy with it?

**Comment by nickcolley** [comment]
If we're committing the the same files that are in https://github.com/alphagov/fonts, should we remove the https://github.com/alphagov/fonts repository?

**Comment by Nooshu** [comment]
The fonts repository can be seen as a store for all versions we have of the font, both hinted and un-hinted. Somewhere teams can reference in the future if they need a copy of the originals.

**Pull request merged**